Maple2.Server.Game/PacketHandlers/MeretMarketHandler.cs

```diff
@@ -0,0 +1,310 @@
+        var gender = packet.Read<GenderFilterFlag>();
+        var job = packet.Read<JobFilterFlag>();
+        var sortBy = packet.Read<MeretMarketSort>();
+        string searchString = packet.ReadUnicodeString();
+        int startPage = packet.ReadInt();
+        int unknown = packet.ReadInt(); // repeats the startPage value
+        MeretMarketSection section = GetMarketSection(packet.ReadByte());
+        packet.ReadByte();
+        byte itemsPerPage = packet.ReadByte();
+
+        ICollection<MarketItem> entries = GetItems(session, section, sortBy, tabId, gender, job, searchString);
+        int totalItems = entries.Count;
+        entries = TakeLimit(entries, startPage, itemsPerPage);
+
+        session.Send(MeretMarketPacket.LoadItems(entries.ToList(), totalItems, startPage));
+    }
+
+    private static void HandlePurchase(GameSession session, IByteReader packet) {
+        byte quantity = packet.ReadByte();
+        int premiumMarketId = packet.ReadInt();
+        long ugcItemId = packet.ReadLong();
+        if (ugcItemId > 0) {
+            // TODO: Handle UGC item purchasing
+            return;
+        }
+
+        packet.ReadInt();
+        int childMarketItemId = packet.ReadInt();
+        long unk1 = packet.ReadLong();
+        int itemIndex = packet.ReadInt();
+        int totalQuantity = packet.ReadInt();
+        int unk2 = packet.ReadInt();
+        byte unk3 = packet.ReadByte();
+        string unk4 = packet.ReadUnicodeString();
+        string unk5 = packet.ReadUnicodeString();
+        long price = packet.ReadLong();
+
+        PremiumMarketItem? entry = session.GetPremiumMarketItem(premiumMarketId, childMarketItemId);
+        if (entry == null) {
+            return;
+        }
+
+        // TODO: Find meret market error packets
+        if (entry.SellBeginTime > DateTime.Now.ToEpochSeconds() || entry.SellEndTime < DateTime.Now.ToEpochSeconds()) {
+            return;
+        }
+
+        if ((entry.RequireMinLevel > 0 && entry.RequireMinLevel > session.Player.Value.Character.Level) ||
+            (entry.RequireMaxLevel > 0 && entry.RequireMaxLevel < session.Player.Value.Character.Level)) {
+            return;
+        }
+
+        if (entry.JobRequirement.Code().All(job => job != session.Player.Value.Character.Job.Code())) {
+            return;
+        }
+
+        if (entry.RequireAchievementId > 0 && session.Achievement.HasAchievement(entry.RequireAchievementId, entry.RequireAchievementRank)) {
+            return;
+        }
+
+        price = entry.SalePrice > 0 ? entry.SalePrice : entry.Price;
+        StringCode payResult = Pay(session, price, entry.CurrencyType);
+        if (payResult != StringCode.s_empty_string) {
+            session.Send(NoticePacket.MessageBox(payResult));
+            return;
+        }
+
+        Item? newItem = session.Item.CreateItem(entry.ItemMetadata.Id, entry.Rarity, entry.Quantity + entry.BonusQuantity);
+        if (newItem == null) {
+            return;
```

Comment:
This should be a fatal error since the player already paid for it.

```cs
Logger.Fatal("Failed to create item {ItemId}, {Rarity}, {Quantity}", entry.ItemMetadata.Id, entry.Rarity, entry.Quantity + entry.BonusQuantity);
throw new InvalidOperationException($"Fatal: ...");
```